libkbfs/block_ops.go

```diff
@@ -83,31 +84,22 @@ func (b *BlockOpsStandard) Get(ctx context.Context, kmd KeyMetadata,
 // GetEncodedSize implements the BlockOps interface for
 // BlockOpsStandard.
 func (b *BlockOpsStandard) GetEncodedSize(ctx context.Context, kmd KeyMetadata,
-	blockPtr BlockPointer) (uint32, error) {
+	blockPtr BlockPointer) (uint32, keybase1.BlockStatus, error) {
 	// Check the journal explicitly first, so we don't get stuck in
 	// the block-fetching queue.
 	if journalBServer, ok := b.config.BlockServer().(journalBlockServer); ok {
 		size, found, err := journalBServer.getBlockSizeFromJournal(
 			kmd.TlfID(), blockPtr.ID)
 		if err != nil {
-			return 0, err
+			return 0, 0, err
 		}
-		if found {
-			return size, nil
+		if found && size > 0 {
+			return size, keybase1.BlockStatus_LIVE, nil
 		}
 	}
 
-	// Otherwise fetch the entire block from the server, since we
-	// can't trust the server to report the size without being able
-	// to verify the BlockID.
-	block := NewCommonBlock()
-	errCh := b.queue.Request(ctx, defaultOnDemandRequestPriority, kmd,
-		blockPtr, block, NoCacheEntry)
-	err := <-errCh
-	if err != nil {
-		return 0, err
-	}
-	return block.GetEncodedSize(), nil
+	return b.config.BlockServer().GetEncodedSize(
+		ctx, kmd.TlfID(), blockPtr.ID, blockPtr.Context)
```

Comment:
Why are we skipping the caches for this? `NoCacheEntry` is about what to do with the block that's returned, not about whether to retrieve from caches.

Additionally, I don't think this is the right approach here--`GetEncodedSize` happens quite often and needs to go through the request de-duplication with other `GetEncodedSize` and `GetBlock` requests. By using the `blockRetrievalQueue`, this de-duping comes for free.

Reply:
The whole point here _is_ to hit the servers, because we need to know the server's current view of the status of the block.  That's not something we can productively cache, as far as I know.

`GetEncodedSize` shouldn't actually happen quite often.  It only happens during real conflict resolution.  Only one of those happens at a time, and it will only be called for each block once during that CR.

I don't understand your point about `NoCacheEntry` or using the retrieval queue -- the queue doesn't support `GetEncodedSize` yet as far as I know, and I don't see much benefit in making it do so.

Reply:
(This no longer fetches the whole block, in case that wasn't clear.  Just the size and the status.)

Reply:
I mean split out the functions, such that we only retrieve the status via RPC when we needed, so that when we only need the size we use the cached value.

Reply:
We also `GetEncodedSize` in `syncAllLocked`: https://github.com/keybase/kbfs/blob/ed63bec510192b0739c519293d0133a6205bd281/libkbfs/folder_branch_ops.go#L4593-L4596

Reply:
Oh, but on the `syncAllLocked` path, [this line](https://github.com/keybase/kbfs/pull/1764/files#diff-63b0a199cafd8f676224371308119991R611) is always going to come out to `false`, and so we _will_ use sizes from the cache because of [this](https://github.com/keybase/kbfs/pull/1764/files#diff-ca9a77b130940ad6ae13c6cb2d59dcd8R289).

We only have to hit the server during actual CR, when the status of the blocks we are unreferencing is ambiguous.

> I mean split out the functions, such that we only retrieve the status via RPC when needed, so that when we only need the size we use the cached value.

That's essentially what the code is doing as explained above -- it uses the cached size whenever the status isn't actually needed. If we split it into two functions, then to avoid two RPCs we need to add a separate cache for the block sizes somewhere (since we won't be downloading the actual block, so it can't go in the `BlockCache`).  That seems like overkill.

Reply:
> Oh, but on the syncAllLocked path, this line is always going to come out to false, and so we will use sizes from the cache because of this.
Cool, I'm happy with this. Thanks for the explanation!

Reply:
Oh actually, if the block isn't cached in memory, you're still skipping the disk cache.